Allow creation transformation functions over data lake variables

rafaellehmkuhl opened the pull request:
This PR adds a powerful new feature to Cockpit: being able to create transformation functions. What that means is that it basically allows the user to create new data lake variables based on existing ones.

Those functions can be as simple as summing two variables, or as complex as the user wants. Want `if` clauses? You have it. The transformation function is basically an `eval` over the expression input.

If the user does not include a `return` statement (as the video example below, where we use `{{ var1 }} + {{ var2 }}`, the system does it automatically for them. The user can have complex functions with their own `returns` as well.

https://github.com/user-attachments/assets/122290fb-9e28-483e-8e75-6fc73e77427c

If the expression fails to be evaluated (e.g.: math error, variable access error, etc), it will try again in 10ms, than 20ms, than 40, doubling on every sequential error, till each reaches 10 seconds. That heuristic was put there to prevent error spamming. 

This functionality will be used with the Radcam joystick variables to create the final zoom and focus values to be sent to the autopilot.

Fix #1642

ArturoManzoli reviewed (changes requested):
Awesome feature!

UI wise is good to change some stuff:
1- The save button is blue (probably as primary color). Change it to our white/transparent standard;
2- The idea of using the placeholder for some quick use guide is very nice! Add some usage examples for boolean and text variables;
3- The placeholder on the description could simply be "optional description" or something more direct. The current placeholder is too descriptive I think;
4- The click-to-copy button isn't working. This is important to have imo;

![image](https://github.com/user-attachments/assets/5673dfb2-b79a-4a34-8930-a46e1404eee2)

Code will be reviewed next

ES-Alexander reviewed (changes requested):
1. It's confusing to have transforming functions in a separate menu to the data lake variable monitor, especially since the transforming functions are just compound variables
    - Can we merge them? I think they can even be in the same table
        - We could add a `Source` column (between `ID` and `Type`) that has icons to specify whether a variable is from MAVLink (logo), Cockpit internal (logo), Other (`mdi-exit-run`), or compound (`mdi-database-settings-outline`)
            - The compound ones could have the edit/delete icons included in the Source column, or we could use `database-edit-outline` as the icon and only allow deleting from inside the editor
            - In future clicking on an Other one could list the Actions and/or widgets that are known to set that variable, and possibly even jump to their configuration page
        - We could add a "New compound variable" button under the table
        - The `ID` column contents seem largely redundant with the Name (only humans are looking at the table)
            - we could regain some table width by removing the text for the IDs and just leaving the copy button
    - More generally, I feel like Joystick, Actions, Data Lake, and MAVLink belong together, separated from the rest of the settings stuff, because they're all related to things you can _do_ with Cockpit - I think they can all go in the "Tools" menu, which would also balance the load with the "Settings" menu
        - It seems reasonable to have a shortcut from the MAVLink page to the Actions page, in case someone wants to manually send a message
2. Having the variables accessible is a a great idea, but it's seemingly quite awkward to use at the moment - could we make it inline instead of a separate form field?
    - I'm thinking there could be an auto-complete dropdown where the cursor is, whenever it's at the end of a zero or more length set of valid ID characters (and searched using those characters), between either a pair of double curly braces, or an open double curly brace and the end of the input (e.g. when first typing)
        - This could support normal clicking to select and replace the currently typed characters with the selected variable's ID, and ideally also tab-completion
    - I would prefer for this to be implemented here, but it can be opened as an Issue and handled later if necessary, since it's polish / quality of life rather than fundamental to the feature

ArturoManzoli reviewed (changes requested):
I have one more suggestion before approval:
On FHD (most of our users) screens (most of our users) the modal height is larger than the screen height, and so a scrollbar is present, making the user roll down the modal to see the save button. 

But this can be solved quite simply by adding a density="compact" and hide-details to all the `<v-text-fields />`, `<v-select />` and `<v-text-area />`. Also increase the gap to 7 on line 8 and invert the margin to `-mt-1` on line 21.

This way we save important vertical space.

Check the result:
**Before**
![image](https://github.com/user-attachments/assets/8752102a-6b17-4554-9ae3-aa0232e2d9f8)

**After**
![image](https://github.com/user-attachments/assets/f4da369e-0c41-4078-9a4a-ee93e9b7ee50)

rafaellehmkuhl commented:
> I have one more suggestion before approval: On FHD screens (most of our users) the modal height is larger than the screen height, and so a scrollbar is present, making the user roll down the modal to see the save button.
> 
> But this can be solved quite simply by adding a density="compact" and hide-details to all the `<v-text-fields />`, `<v-select />` and `<v-text-area />`. Also increase the gap to 7 on line 8 and invert the margin to `-mt-1` on line 21.
> 
> This way we save important vertical space.
> 

Good catch! Fixed.

ArturoManzoli reviewed (approved):
Nice feature. Will have a many uses over time

ES-Alexander reviewed (changes requested):
Nice progress! :D

Some suggestions/notes:
1. Can the ID column be moved to the left of the name? It seems like it could take up much less space
2. Comments are included in the placeholder, but using a comment currently stops the expression from evaluating (with no notice/warnings)
    - Ideally comments would be allowed, and would just get stripped out before being passed to the evaluation function
3. The "Description" field seems kind of redundant - could we just automatically extract a description using a comment on the first line, if there is one, and otherwise assume there's no description?
    - Longer term it would be nice to include automatic descriptions for the MAVLink fields and the like, using metadata, but that can be raised in an Issue
5. Could we have a single row with `[Variable Name ] -> [Variable ID ]`? I don't think names are expected (or encouraged) to be super long
6. If we don't expect the available variable types to change in the near future, could we replace the "Variable Type" dropdown with a row of radio-buttons, so it's a single click to change, and uses the space better?
7. Editing the ID is disabled for existing variables, so maybe the icon could change to a copy one (so it can actually be used for something)?
8. Compound variables don't get initialised on startup
9. It would be nice to be able to delete variables that were created using Input Widgets and the like, but that can be raised as an Issue
10. The new +/- buttons on the Dial don't actually update the underlying data lake variable (can be an Issue, I can open it later - short on time right now)

rafaellehmkuhl commented on src/components/TransformingFunctionDialog.vue:
```diff
@@ -0,0 +1,396 @@
+            />
+            <div class="text-xs text-gray-400">
+              Create complex transformations by combining existing Data Lake variables using JavaScript expressions.
+              Type '\{\{' to access available variables. Return is optional, but should be included in complex
```
Fixed. 
This actually doesn't work, and we need to use `&#123;&#123;`.

rafaellehmkuhl commented:
> Nice progress, and already much nicer to use! :D
> 
> Some suggestions/notes:
> 
> 1. Can the ID column be moved to the left of the name? It seems like it could take up much less space
> 2. Comments are included in the placeholder, but using a comment currently stops the expression from evaluating (with no notice/warnings)
>    
>    * Ideally comments would be allowed, and would just get stripped out before being passed to the evaluation function
> 3. The "Description" field seems kind of redundant - could we just automatically extract a description using a comment on the first line, if there is one, and otherwise assume there's no description?
>    
>    * Longer term it would be nice to include automatic descriptions for the MAVLink fields and the like, using metadata, but that can be raised in an Issue
> 4. Could we have a single row with `[Variable Name ] -> [Variable ID ]`? I don't think names are expected (or encouraged) to be super long
> 5. If we don't expect the available variable types to change in the near future, could we replace the "Variable Type" dropdown with a row of radio-buttons, so it's a single click to change, and uses the space better?
> 6. Editing the ID is disabled for existing variables, so maybe the icon could change to a copy one (so it can actually be used for something)?
> 7. Compound variables don't get initialised on startup
>    
>    * This should happen after the internal variables get initialised
> 8. It would be nice to be able to delete variables that were created using Input Widgets and the like, but that can be raised as an Issue
> 9. The new +/- buttons on the Dial don't actually update the underlying data lake variable (can be an Issue, I can open it later - short on time right now)

1 & 4. Tried here, but it gets very strange. I don't think it's a big deal, specially since we already reduced the vertical space usage by a lot in the last change. It fits in both a FHD screen as well as a MacAir one.
2. This was actually a bug caused by my automatic addition of the `return` keyword on expressions missing it. Just fixed it. It now supports multi-line comments as well.
3. Actually the description is something available for all data-lake variables. I've replaced the text-field with a button that the user should click if they want to add a description, so we save some vertical space by default.
5. Done.
6. I prefer not to do that. They already have the copy functionality elsewhere.
7. Nice catch. Fixed it!
8. Agree that it's useful, but out of scope here.
9. It's a bug in the Dial widget. Should be fixed there as well.

ES-Alexander commented:
> Tried here, but it gets very strange. I don't think it's a big deal, specially since we already reduced the vertical space usage by a lot in the last change. It fits in both a FHD screen as well as a MacAir one.

Not sure what you're referring to here - I was asking about the ID column in the table, which is just copy icons, but currently is somewhat lost in space/padding between the names and types. Moving it to before the names seems like we could reduce the padding on the right and save a bunch of horizontal space, while also making it clearer that it's copying something related to the names :-)

rafaellehmkuhl commented:
> > Tried here, but it gets very strange. I don't think it's a big deal, specially since we already reduced the vertical space usage by a lot in the last change. It fits in both a FHD screen as well as a MacAir one.
> 
> 
> 
> Not sure what you're referring to here - I was asking about the ID column in the table, which is just copy icons, but currently is somewhat lost in space/padding between the names and types. Moving it to before the names seems like we could reduce the padding on the right and save a bunch of horizontal space, while also making it clearer that it's copying something related to the names :-)

Oh, I thought you mentioned the ID input in the dialog.

The ID column in the table was removed, like suggested in a previous review.

ES-Alexander commented:
> The ID column in the table was removed, like suggested in a previous review.

Lol, it still exists (as intended) - it just only contains copy icons now. I'm suggesting to move those copy icons to _before_ the Name column, instead of after it:
<img width="531" alt="Screenshot 2025-02-27 at 11 13 33 am" src="https://github.com/user-attachments/assets/20fd4503-0e3f-4dda-af1d-783d2a7223f8" />

(ignore the background-colour changes from moving parts of a screenshot around)

ES-Alexander reviewed (dismissed):
> The ID column in the table was removed, like suggested in a previous review.

1. Lol, it still exists (as intended) - it just only contains copy icons now. I'm suggesting to move those copy icons to _before_ the Name column, instead of after it:
    <img width="531" alt="Screenshot 2025-02-27 at 11 13 33 am" src="https://github.com/user-attachments/assets/20fd4503-0e3f-4dda-af1d-783d2a7223f8" />
    (ignore the background-colour changes from moving parts of a screenshot around)

> This was actually a bug caused by my automatic addition of the return keyword on expressions missing it. Just fixed it. It now supports multi-line comments as well.

2. Tried this, and can confirm it now works for single line `//` comments, and multiline `/**/` comments 👍 
    
    It does still fail on `/**/` comments that are confined to a single line 😕
That's not a big problem, but is an unexpected failure case. I can raise an issue if you don't think that's worth fixing here.

>  ... we already reduced the vertical space usage by a lot in the last change. It fits in both a FHD screen as well as a MacAir one.
> I've replaced the text-field with a button that the user should click if they want to add a description, so we save some vertical space by default.

3. I use a slightly scaled display on my 16" MBP, so it has a more comfortable (larger) text size:
    <img width="461" alt="Screenshot 2025-02-27 at 11 25 08 am" src="https://github.com/user-attachments/assets/fea4a595-f8bc-4b11-8ba1-d626b7b9b688" />
    
    I have to vertically scroll on the menu to be able to access the save/cancel buttons, even with the "Description" changed to a button:
    <img width="599" alt="Screenshot 2025-02-27 at 11 31 06 am" src="https://github.com/user-attachments/assets/d4d1fb1c-5fff-4b41-8f35-ee0cded43ea4" />
    
    Outside of vertical space though, my suggestion to extract it from the first line comment in the code was also because we already show code comments as possible (in the placeholder text examples), so it seems intuitive to include some kind of description there, in which case it seems unintuitive to _also_ have a description specified in a separate form field underneath the code window. Just an opinion 🤷 

> ... the description is something available for all data-lake variables.

3. (part 2) Is it expected to be populated for the MAVLink message fields? Because it's not at the moment:
    <img width="405" alt="Screenshot 2025-02-27 at 11 43 53 am" src="https://github.com/user-attachments/assets/af8a664e-c70c-4ca6-a9a6-a0b92fd6dfe9" />
    
    As mentioned in my previous review, this can be raised in an issue if we don't want to handle it here :-)

>> 4. Could we have a single row with [Variable Name ] -> [Variable ID ]? I don't think names are expected (or encouraged) to be super long
>
> Same as 1.

4. Per 3, it may still be worth reducing vertical space usage. This suggestion was also ID to the right of the name, whereas you previously thought 1 was suggesting ID to the left of the name. Not a big deal, but may be worth a try if you haven't tested that option 🤷 

5. Radio buttons look great - thanks! :-)

6. No worries :-)

7. Nice!

8. Sure (#1721) :-)

9. Sure (#1720) :-)

---

10. New bug found: if you edit a compound variable, press "cancel", then edit it again, it opens as an edit pane with new widget form values, and cannot be saved. It's possible to get back to the real edit pane by editing another variable first, or by closing and reopening the data lake monitor, but that shouldn't be required.
    
    https://github.com/user-attachments/assets/5a317d3e-02ef-41ab-b0c4-3de63cbce248

rafaellehmkuhl commented:
> > The ID column in the table was removed, like suggested in a previous review.
> 
> 1. Lol, it still exists (as intended) - it just only contains copy icons now. I'm suggesting to move those copy icons to _before_ the Name column, instead of after it:

Done.

> > This was actually a bug caused by my automatic addition of the return keyword on expressions missing it. Just fixed it. It now supports multi-line comments as well.
> 
> 2. Tried this, and can confirm it now works for single line `//` comments, and multiline `/**/` comments 👍
>    It does still fail on `/**/` comments that are confined to a single line 😕
>    That's not a big problem, but is an unexpected failure case. I can raise an issue if you don't think that's worth fixing here.

Fixed this case as well.

> > ... we already reduced the vertical space usage by a lot in the last change. It fits in both a FHD screen as well as a MacAir one.
> > I've replaced the text-field with a button that the user should click if they want to add a description, so we save some vertical space by default.
> 
> 3. I use a slightly scaled display on my 16" MBP, so it has a more comfortable (larger) text size:
>    I have to vertically scroll on the menu to be able to access the save/cancel buttons, even with the "Description" changed to a button:
>    Outside of vertical space though, my suggestion to extract it from the first line comment in the code was also because we already show code comments as possible (in the placeholder text examples), so it seems intuitive to include some kind of description there, in which case it seems unintuitive to _also_ have a description specified in a separate form field underneath the code window. Just an opinion 🤷

Removed the fixed description and added an info button. Gained a lot of vertical space indeed.

> > ... the description is something available for all data-lake variables.
> 
> 3. (part 2) Is it expected to be populated for the MAVLink message fields? Because it's not at the moment:
>    As mentioned in my previous review, this can be raised in an issue if we don't want to handle it here :-)

The MAVLink variables are indeed not expected to have descriptions, as they are automatically generated and no metadata is consulted to add it.

> 4. Per 3, it may still be worth reducing vertical space usage. This suggestion was also ID to the right of the name, whereas you previously thought 1 was suggesting ID to the left of the name. Not a big deal, but may be worth a try if you haven't tested that option 🤷

Done.

> 5. Radio buttons look great - thanks! :-)

Thanks :D

> 8. Sure ([frontend: data lake: allow deleting variables in the monitor #1721](https://github.com/bluerobotics/cockpit/issues/1721)) :-)
> 9. Sure ([Dial: stop resizing value indicator with contents #1720](https://github.com/bluerobotics/cockpit/issues/1720)) :-)

Thanks!

> 10. New bug found: if you edit a compound variable, press "cancel", then edit it again, it opens as an edit pane with new widget form values, and cannot be saved. It's possible to get back to the real edit pane by editing another variable first, or by closing and reopening the data lake monitor, but that shouldn't be required.

I had a hard time reproducing it, but I think I've fixed it now. Can you check again?

ArturoManzoli reviewed (changes requested):
Looks great. It's almost approved.

(1) Just take out the `mt-1` from the mdi-pencil;

(2) And in my opinion, the description always set to 1 row is fine. It's one less click for the user to add a description, and if it's a really big description, the user can pull down the text area and enlarge it manually.


![image](https://github.com/user-attachments/assets/4d0a758c-d871-4c86-9843-74a9fa357ea4)

rafaellehmkuhl commented:
> Looks great. It's almost approved.
> 
> (1) Just take out the `mt-1` from the mdi-pencil;

Inverted to `-mt-1` so it doesn't take extra vertical space and also gets aligned with the text input.

> (2) And in my opinion, the description always set to 1 row is fine. It's one less click for the user to add a description, and if it's a really big description, the user can pull down the text area and enlarge it manually.

Good idea. Done.

ArturoManzoli reviewed (approved):
Nice!
